### ISSUE TITLE
feature: Protect volumes against deletion

### ISSUE THREAD

**Issue by oliv3r** [moby/moby#45725]
### Description

Docker mantra is 'containers are **ephemeral**'. Good, fair, I agree. But more then often, pet-servers and desktops become a desert of containers and volumes (yeah, I know everybody is aware of this :p).

There's of course `docker prune --all --force --volumes` that will clean up all that cruft! Great. However there is one big issue, persistent storage volumes. E.g. the volume that holds the database, the volume that contains those certbot certificates etc.

And yes, they can be restored from a backup, but it's not impossible to assume, that people are very bad at creating backups (especially since we can't even copy volumes with the volume tooling easily #33782). But the fear of loosing a volume, where all important data is on, is real, and 'just make sure you backup everything first' is good mantra, but this shouldn't be the only way to protect things.

So here is my proposal, which deals only with volumes, unllike #9878 #7523, which deals with containers as a whole, and goes beyond anonymous volumes as mentioned here #44216  (we really should have a `docker volume update` command to be able to add labels ... #21721.

Add the `volume create` flag (and label via update)
```json
"com.docker.volume.protected": "",
```
(the name is not important).

Whenever any docker rm action touches this volume (be it prune, or manual rm), it will cause a question 'this is a protected volume, do you really want to remove this protected volume', even if the `--force` flag is used. It requires user confirmation. A new flag is needed of course `--force-protected` to bypass this question without interaction, as the normal `--force` flag will only ask if the operation as a whole should start (system prune).

If there IS content in the label, it should be used as a password. Not because it is super secret, but it adds an additional layer of protection on the removal. `--force-protected <string>` is thus needed for scripting purposes.

This is of course very heavy on the paranoya, but removing a data volume, that wasn't supposed to be pruned, can be very costly ($$) even if there's backups. A lot of docker users would sleep much safer knowing their persistent volumes are truly persistent.

Compose could support all of the above, as the keyword is not intended to be secret.

```yaml
volumes:
  mypersitent_volume:
   persistent: true
  myother_volume:
   persistent: keyword
```
The above example does entail that 'true, yes, no, false' would be words not possible as keywords.

Though having `labels` available on the volume level to add from the compose file  https://github.com/docker/compose/issues/10689 ;)

TLDR;

```yaml
volumes:
  mypersitent_volume:
   persistent: true
  myother_volume:
   persistent: keyword
```

```json
"Labels": {
  "com.docker.volume.protected": "keyword",
},
```

```
Options:
  -f, --force   Force the removal of one or more volumes
  -p,--force-prot Force the removal of protected volumes
  -k,--keyword  Protected volume keyword
```
Afaik we can't do `--force-prot [string]` as that confuses getopts for sure. And this won't work with multiple volumes (unless the keyword is the same).

This can probably be expanded to all others too (networks, containers).

P.S. Yes, we can do an ugly work-around such as `docker run --rm -it -v myvol:/myvol -v myvol1:/myvol1 /bin/sh` to make sure these volumes are in use, while everything is 'shut down'. But its a crudge to say the least.

**Comment by cpuguy83** [comment]
Volumes that have been explicitly named are no longer considered for pruning unless (without an extra param).
It seems like that would satisfy your case?

**Comment by oliv3r** [comment]
@cpuguy83 I think if you prune with `docker system prune --all --volume --yes` as a 'general clean everything up' it will happily remove your named volumes. We want 'protection' of our data volumes, without having to go through a cycle of 'prune all, and then restore backups'.

**Comment by neersighted** [comment]
`docker system prune --all --volume` does not prune named volumes; only `docker volume prune -a` does, and it explicitly states what it will do, and requires confirmation.

I'm going to close this as already implemented, but please do clarify if you think there is more work to do.

**Comment by oliv3r** [comment]
@neersighted thanks for the response, sorry for waiting for so long to 
reply/reopen. but I don't think we understood eachother entirely/share 
concerns. 

The distinction between named and anonymous volumes addresses accidental 
pruning, but it does not protect against accidental explicit deletion or 
muscle-memory bypasses (like habit-clicking through `[y/N]` prompts or using 
`-f`). 

For high-value, manually created persistent data—like a primary password 
database—we need an explicit "soft lock" or protection mechanism. 

The goal here is a GitHub-style destruction prevention workflow:

1. **Bulk actions (`prune`):** The volume is completely **skipped** (with a 
   warning), even if `--all` or `--force` is used. This ensures automated 
   cleanup scripts don't inadvertently wipe critical data.
2. **Direct actions (`rm`):** A targeted `docker volume rm <name>` must 
   **fail** outright by default. Deletion should require a secondary, 
   intentional verification step where the user must explicitly type out the 
   volume name to confirm destruction (e.g., 
   `docker volume rm --confirm-name=<name> <name>`).

This introduces necessary friction that cannot be bypassed by a typo or terminal 
muscle memory, ensuring critical data is actually safe.

Gemini produced the following; I dind't read it, but just a 'ohh this is what you mean' kind of thought...

```patch
diff --git a/cli/command/volume/remove.go b/cli/command/volume/remove.go
index z9y8x7w..v6u5t4s 100644
--- a/cli/command/volume/remove.go
+++ b/cli/command/volume/remove.go
@@ -22,6 +22,7 @@ type removeOptions struct {
 	force   bool
+	confirm string
 }
 
 func NewRemoveCommand(dockerCli command.Cli) *cobra.Command {
@@ -34,6 +35,7 @@ func NewRemoveCommand(dockerCli command.Cli) *cobra.Command {
 	
 	flags := cmd.Flags()
 	flags.BoolVarP(&opts.force, "force", "f", false, "Force the removal of one or more volumes")
+	flags.StringVar(&opts.confirm, "confirm-name", "", "Explicitly type the volume name to confirm destruction of a protected volume")
 	return cmd
 }
 
@@ -50,6 +52,13 @@ func runRemove(ctx context.Context, dockerCli command.Cli, opts removeOptions, v
 		if err != nil {
 			return err
 		}
+
+		// Protect high-value targets from muscle-memory deletion
+		if vol.IsProtected() && opts.confirm != vol.Name() {
+			return fmt.Errorf("Volume %s is protected. To delete, you must pass the exact name via: --confirm-name=%s", vol.Name(), vol.Name())
+		}
+
 		if err := client.VolumeRemove(ctx, vol.Name(), opts.force); err != nil {
 			return err
 		}
diff --git a/daemon/cluster/executor/container/volume.go b/daemon/cluster/executor/container/volume.go
index a1b2c3d..e4f5g6h 100644
--- a/daemon/cluster/executor/container/volume.go
+++ b/daemon/cluster/executor/container/volume.go
@@ -104,6 +104,11 @@ func (d *Daemon) VolumesPrune(ctx context.Context, pruneFilters filters.Args) (*
 		if err != nil {
 			return nil, err
 		}
+
+		// Skip protected volumes entirely during bulk prune operations
+		if v.IsProtected() {
+			logrus.Warnf("Skipping protected volume: %s", v.Name())
+			continue
+		}
+
 		if IsInUse(v) {
 			continue
 		}
diff --git a/daemon/volume.go b/daemon/volume.go
index x9y8x7w..v6u5t4s 100644
--- a/daemon/volume.go
+++ b/daemon/volume.go
@@ -45,3 +45,17 @@ func (daemon *Daemon) VolumeUpdate(ctx context.Context, name string, labels map[
 	if err != nil {
 		return err
 	}
+
+	// Update labels in the volume database/metadata file
+	for k, v := range labels {
+		vol.Labels()[k] = v
+	}
+
+	return daemon.volumes.Save(vol)
 }
diff --git a/volume/store/store.go b/volume/store/store.go
index j8k9l0m..n1o2p3q 100644
--- a/volume/store/store.go
+++ b/volume/store/store.go
@@ -112,6 +112,11 @@ func (s *Store) Create(ctx context.Context, name, driver string, opts ...VolumeO
 	if err != nil {
 		return nil, err
 	}
+
+	if v.IsProtected() {
+		logrus.Infof("Volume %s created with destruction protection active.", v.Name())
+	}
+
 	return v, nil
 }
diff --git a/volume/volume.go b/volume/volume.go
index b1c2a3d..e4f5g6h 100644
--- a/volume/volume.go
+++ b/volume/volume.go
@@ -5,6 +5,14 @@ const (
+	// LabelProtected specifies that a volume cannot be deleted without explicit confirmation
+	LabelProtected = "com.docker.volume.protected"
 )
 
+func (v *LocalVolume) IsProtected() bool {
+	if v.labels != nil && v.labels[LabelProtected] == "true" {
+		return true
+	}
+	return false
+}
```
